src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java

```diff
@@ -0,0 +1,147 @@
+package com.atlassian.bitbucket.jenkins.internal.client;
+
+import com.atlassian.bitbucket.jenkins.internal.client.exception.BitbucketClientException;
+import com.atlassian.bitbucket.jenkins.internal.client.exception.NoContentException;
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketPage;
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketResponse;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import okhttp3.HttpUrl;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import static java.util.Objects.requireNonNull;
+import static okhttp3.HttpUrl.parse;
+
+public class BitbucketRequestExecutor {
+
+    private static final String API_VERSION = "1.0";
+    private static final Logger log = Logger.getLogger(BitbucketRequestExecutor.class);
+
+    private final HttpUrl bitbucketBaseUrl;
+    private final HttpUrl coreRestPathUrl;
+    private final HttpRequestExecutor httpRequestExecutor;
+    private final ObjectMapper objectMapper;
+    private final BitbucketCredentials credentials;
+
+    public BitbucketRequestExecutor(String bitbucketBaseUrl,
+                                    HttpRequestExecutor httpRequestExecutor, ObjectMapper objectMapper,
+                                    BitbucketCredentials credentials) {
```

Comment:
> BitbucketCredentialsAdaptor.createWithFallback() is redundant
I don't agree. The purpose of having this is to make explicit that we are using admin credential as a fallback. To me, it helps in readability if you compare it against simple assigning it to BitbuckerServerConfiguration credential if it is null. Also, the decision of which credentials to use should be with Configurations. 

> I do understand your goal of keeping BitbucketClientFactory independent of Jenkins and this can be achieved by changing ClientFactoryProvider.getClient() to accept Jenkins Credentials
I don't agree. They need to be kept separate. it is a classic example of dependency inversion. Source code dependency of the client module should not be on Jenkins classes. This helps in creating pluggable architecture. Having Jenkins Credential in client means the dependency arrow is bidirectional which should be avoided.

Reply:
> I don't agree. The purpose of having this is to make explicit that we are using admin credential as a fallback. To me, it helps in readability if you compare it against simple assigning it to BitbuckerServerConfiguration credential if it is null. Also, the decision of which credentials to use should be with Configurations

I agree that the fallback should be present at the caller location, but every caller shouldn't require to convert Jenkins credentials to Bitbucket credentials because they don't really care about it.

> I don't agree. They need to be kept separate. it is a classic example of dependency inversion. Source code dependency of the client module should not be on Jenkins classes. This helps in creating pluggable architecture. Having Jenkins Credential in client means the dependency arrow is bidirectional which should be avoided.

As you added a `BitbucketRequestExecutor` which takes care of JSON processing, you could add a wrapper on top of `ClientFactoryProvider` to pass in Jenkins credentials. The end goal is that every user of client doesn't have to convert Jenkins credentials to `BitbucketCredential` because all consumers of Bitbucket client will have Jenkins credentials. API should be designed to make the life of consumer easy. I am worried that we are over engineering stuff.

Reply:
I fail to understand the concern. There is `BitbucketCredentialsAdaptor` which converts Jenkins credentials to Bitbucket Credentials. Previously, every configuration used to pass Credentials and BitbucketServerConfiguration to Client Provider which not only used to initialize HTTP stuff but also decided which credentials to use. I  broke this coupling.